### DISCUSSION TITLE
The current Bazel uses Starlark to write the rules.   Why not just use JAVA to write the rules directly?

### DISCUSSION THREAD

**Discussion by vaneyli** [bazelbuild/bazel#23416] [Ideas]
Just use JAVA to define all rules.   I think that will omit the process of parsering.  All are Java.  Thanks

**Comment by katre** [comment]
We actually used to write all of the rule logic in Java, but migrated to using Starlark to solve a lot of problems. A short list, from memory:
- The native Java rules (many of which still exist) are part of the Bazel binary, and can't be updated separately
  - Being able to release and iterate on rules separately from the now-annual Bazel release is a huge benefit to users
- The Starlark rules have a well-defined and limited API available
  - Java rules frequently used to accidentally call into unoptimized libraries and cause poor performance at scale
  - Java rules caused the rule API to expand, because it was so easy to just add a new layer of indirection

Because of the benefits, we won't be going back to Java rules, but we definitely want to improve the Starlark performance. If you have any suggestions feel free to reply or open a feature request!
